openedx/core/djangoapps/content/block_structure/models.py

```diff
@@ -59,6 +61,63 @@ def _bs_model_storage():
     )
 
 
+class CustomizableFileField(models.FileField):
+    """
+    Subclass of FileField that allows custom settings to not
+    be serialized (hard-coded) in migrations. Otherwise,
+    migrations include optional settings for storage (such as
+    the storage class and bucket name); we don't want to
+    create new migration files for each configuration change.
+    """
+    def __init__(self, *args, **kwargs):
+        kwargs.update(dict(
+            upload_to=_path_name,
+            storage=_bs_model_storage(),
+            max_length=500,  # allocate enough for base path + prefix + usage_key + timestamp in filepath
+        ))
+        super(CustomizableFileField, self).__init__(*args, **kwargs)
+
+    def deconstruct(self):
+        name, path, args, kwargs = super(CustomizableFileField, self).deconstruct()
+        del kwargs['upload_to']
+        del kwargs['storage']
+        del kwargs['max_length']
+        return name, path, args, kwargs
+
+
+@contextmanager
+def _storage_error_handling(bs_model, operation, is_read_operation=False):
```

Comment:
Thanks @ormsbee!

2. My main concern was on reads, more than writes.  For writes, yes, we have celery tasks - and, (in the future) we can use the `PersistOnFailureTask` to make sure we track and retry all failed tasks once S3 is back up.  However, for reads, an issue with S3 could potentially halt the LMS.  I'm going to merge this PR as is, for now.  But would like to see if we can do something better here.

3. Yes.  We're doing this now, with the new call to [transaction.atomic](https://github.com/edx/edx-platform/pull/14717/files#diff-b3d6ed3ee8b1435a3bd338b0e45dc0b0R198).

4. Yes.  We do:  `generate_course_blocks` allows us to re-generate block-structures.  Sounds good on not worrying about this in the short-term, for now.

5. Agree fully.